pg_rational.c

```diff
@@ -111,32 +111,47 @@ rational_in_float(PG_FUNCTION_ARGS)
 {
 	float8		target = PG_GETARG_FLOAT8(0),
 				z,
-				prev_denom,
+				fnumer,
+				fdenom,
 				error;
 	int32		temp,
+				prev_denom,
 				sign;
 	Rational   *result = palloc(sizeof(Rational));
 
-	if (target == floor(target))
+	if (target == (int32)target)
 	{
-		result->numer = floor(target);
+		result->numer = (int32)target;
 		result->denom = 1;
 		PG_RETURN_POINTER(result);
 	}
 
 	sign = target < 0.0 ? -1 : 1;
 	target = fabs(target);
 
+	if (target > INT32_MAX) {
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("value too large for rational")));
+	}
 	z = target;
 	prev_denom = 0;
+	result->numer = (int32)round(target);
 	result->denom = 1;
 	do
 	{
 		z = 1.0 / (z - floor(z));
-		temp = result->denom;
-		result->denom = result->denom * floor(z) + prev_denom;
-		prev_denom = temp;
-		result->numer = round(target * result->denom);
+		fdenom = result->denom * floor(z) + prev_denom;
+		if (fdenom > INT32_MAX) {
```

Comment:
What if you did the assignments together, and checked the ranges all at once, like

```c
fdenom = result->denom * floor(z) + prev_denom;
fnumer = round(target * fdenom);
if (fnumer > INT32_MAX || fdenom > INT32_MAX)
	break;
```

Also when we break out of the loop because one of those are too big, should that be considered an error? Looks like it currently just proceeds to return the result.